solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java

```diff
@@ -146,5 +171,100 @@ public float getMaxScore(int upTo) throws IOException {
         return Float.POSITIVE_INFINITY;
       }
     }
+
+    /**
+     * A FeatureScorer that reads the docValues for a field
+     */
+    public class DocValuesFieldValueFeatureScorer extends FeatureWeight.FeatureScorer {
+      final LeafReaderContext context;
+      final DocIdSetIterator docValues;
```

Comment:
Sadly, the `DocValuesIterator` is not public, so I cannot use it here.
But I am working on a restructuring.

Reply:
The restructuring happened in ec4cbfb.
I am not quite sure, whether I like the duplication in lines 216 & 219, but by reversing the check for the type and the check for the presence of the docId, the cast to the correct docValuesType can happen earlier, allowing the use of `advanceExact()`.

`advanceExact()` generally seems to be preferred over `advance()`, can you tell me why?

Reply:
> ... `advanceExact()` generally seems to be preferred over `advance()`, can you tell me why?

Good question, I don't know!

Intuitively something that returns `boolean` rather than `int` when the `int` is not needed seems neater but really we are after performance and not necessarily neatness.

`NumericDocValues.longValue()` mentioned `advanceExact()` so that could be a clue but not an explanation.

Slightly random code browsing to (say) https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.8.2/lucene/core/src/java/org/apache/lucene/codecs/lucene80/Lucene80DocValuesProducer.java finds some delegation to https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.8.2/lucene/core/src/java/org/apache/lucene/codecs/lucene80/IndexedDISI.java and its advanceNext method and mentions of `IndexedDISI` always make me a bit _dizzy_ since I don't yet know much about it.

The last commit on https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.8.2/lucene/core/src/java/org/apache/lucene/index/NumericDocValues.java reads "LUCENE-7462: Give doc values APIs an advanceExact method." and that commit and https://issues.apache.org/jira/browse/LUCENE-7462 should have more info then.

Reply:
Thank you for the links. :)
I only had some time to skim trough it, but this really looks interesting. As soon as my time allows it, I will definitely dig deeper into that.